### ISSUE TITLE
Clear semantics for indicating batch shape for sampling and scoring

### ISSUE THREAD

**Issue by neerajprad** [pyro-ppl/pyro#327]
Currently, one place where the object and functional forms for distributions return different results is in sampling. Distributions take in a `batch_size` parameter in their constructor, which is used to expand the distribution parameters along dimension 0 if `batch_size` is specified. e.g.

```
            if alpha.dim() == 1 and beta.dim() == 1 and batch_size is not None:
                self.alpha = alpha.expand(batch_size, alpha.size(0))
                self.beta = beta.expand(batch_size, beta.size(0))
```

Calling the `sample` method on a distribution then results in a batch of samples since the parameters have been appropriately expanded. We rely on the user to do this expansion of the distribution parameters when we call `sample` using the functional form of the distributions. For scoring, we use the data shape to appropriately shape the distribution parameters, and this is consistent between the two forms. 

As we implement #323, we will need to carefully think about how we incorporate `batch_size`, or more generally `batch_shape` for scoring and sampling. They will likely not be the same for both methods.

 - One way would be for the objects to take in an explicit `batch_shape` argument. If not specified, this is computed using the shape of the distribution parameters, as is implemented now. This will fix the shape of both the samples as well as the return shape of `batch_log_pdf`. This will however require certain changes to the way we are calling distributions right now in our example code. Another issue is that we may not know what the `batch_shape` may be at the time of creating the distribution object.
 - Another approach would be for `batch_shape` and `shape` methods to take in an additional parameter `x` (data). These would then be called at run-time to determine how the attributes of the distribution class need to be scaled for scoring (`batch_log_pdf`). Currently, we scale the distribution parameters using the data, and then call these methods using the scaled parameters. This will not be possible (or rather, desirable) after #323 is implemented since `batch_shape` would use the distribution parameters specified during construction. This seems like a more flexible approach, that will not need changes to our example code. Scoring will work as it does right now for the object form.
 - Deprecate `batch_shape` / `batch_size` entirely. The onus is on the user to specify appropriately resized distribution parameters for scoring and sampling. Simplest to support, but will move all boiler-plate code to clients. Will also require changing our example code.

See also #146.
